### PR TITLE
[WIP] ao_aaudio: new audio output driver for Android O+

### PR DESCRIPTION
**Pull request by tomty89**
Essentially AAudio is the revamped version of OpenSLES, i.e. C API (no JVM/JNI hassle) that matches with the modern day AudioTrack Java API.

As you can see from the code it's much simpler and cleaner than OpenSLES. One big advantage is it finally provides means to query sample rate and buffer size info as per the current audio sink.

Apparently by design it allows us to be notified of audio device disconnection, which may allow us to reinit the ao driver to update it with the aforementioned info whenever there's a device change. However I have literally no knowledge at all about threads (and only a little about mpv's ao infrastructure), which appears to be required to implement it. (Check out about its error callback here: https://developer.android.com/ndk/guides/audio/aaudio/aaudio)

The downside is it doesn't appear to handle device reconnection out of the box like OpenSLES and Java/JNI AudioTrack does. That's the main reason I mark it as WIP. It works well enough for me as long as I stick with the same audio device during playback though.

I used `ao_opensles.c` as the template of it, which may or may not be the optimal way of implementing it (e.g. pull v.s. push? which I don't understand). Also I removed the mutex lock thing (again, threads) as I have no idea what that does and whether it is necessary here. These are the other reasons I marked it as WIP.

### CODE REVIEW

**Comment by tomty89** [comment]
So I found this `ao-reload` command and here's a proof-of-concept:
```
[cplayer] Starting playback...
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [out] 48000Hz stereo 2ch floatp
[af]   [ao] 48000Hz stereo 2ch floatp
[ao] Trying audio driver 'aaudio'
[ao/aaudio] requested format: 48000 Hz, stereo channels, floatp
[ao/aaudio] device buffer: 1536 samples.
[ao/aaudio] using soft-buffer of 9600 samples.
[cplayer] AO: [aaudio] 48000Hz stereo 2ch float
[cplayer] AO: Description: AAudio output
[af] Adding filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float [a]
[af]   [out] 48000Hz stereo 2ch float
[af]   [ao] 48000Hz stereo 2ch float
[cplayer] starting audio playback
[cplayer] playback restart complete
[statusline] A: 00:00:12 / 02:28:43 (0%)

[cplayer] Run command: ao-reload, flags=9, args=[]
[af] Removing filter lavrresample
[af] Adding filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float [a]
[af]   [out] 48000Hz stereo 2ch float
[af]   [ao] 48000Hz stereo 2ch float
[af] Removing filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [out] 48000Hz stereo 2ch floatp
[af]   [ao] 48000Hz stereo 2ch floatp
[ao] Trying audio driver 'aaudio'
[ao/aaudio] requested format: 48000 Hz, stereo channels, floatp
[ao/aaudio] device buffer: 11544 samples.
[ao/aaudio] using soft-buffer of 11544 samples.
[cplayer] AO: [aaudio] 48000Hz stereo 2ch float
[cplayer] AO: Description: AAudio output
[af] Adding filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float [a]
[af]   [out] 48000Hz stereo 2ch float
[af]   [ao] 48000Hz stereo 2ch float
[statusline] A: 00:00:24 / 02:28:43 (0%)

[cplayer] Run command: ao-reload, flags=9, args=[]
[af] Removing filter lavrresample
[af] Adding filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float [a]
[af]   [out] 48000Hz stereo 2ch float
[af]   [ao] 48000Hz stereo 2ch float
[af] Removing filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [out] 48000Hz stereo 2ch floatp
[af]   [ao] 48000Hz stereo 2ch floatp
[ao] Trying audio driver 'aaudio'
[ao/aaudio] requested format: 48000 Hz, stereo channels, floatp
[ao/aaudio] device buffer: 1536 samples.
[ao/aaudio] using soft-buffer of 9600 samples.
[cplayer] AO: [aaudio] 48000Hz stereo 2ch float
[cplayer] AO: Description: AAudio output
[af] Adding filter lavrresample
[af] Audio filter chain:
[af]   [in] 48000Hz stereo 2ch floatp
[af]   [lavrresample] 48000Hz stereo 2ch float [a]
[af]   [out] 48000Hz stereo 2ch float
[af]   [ao] 48000Hz stereo 2ch float
[statusline] A: 00:00:26 / 02:28:43 (0%)
```

As you can see the sample rate and device buffer size being updated after switching to and from a Bluetooth audio device, so that we will not get buffer underrun when we are on Bluetooth, yet gets back to low latency when we are back on wired. This practically allows us to set `audio-buffer` to `0`.

Though for some reason mpv segfaults when I switch away from Termux (the terminal app/prefixed environment I run mpv from) after an `ao-reload` (it does not segfault util I switch away or if the command was never run). Got to fix that out even if we have means to trigger `ao-reload` in the driver.

I figured out how to make the ao at least work like `ao_opensles` when there's s change of device though. It's just without a full reload the playback will not be in optimal state. Still better than `ao_opensles` though, as `numFrames` is dynamic.

**Comment by tomty89** [comment]
Hmm the weird segfault after `ao-reload` only occurs with `ao_aaudio` (even when there's no device change) but not `ao_opensles`. Thought it might have something to do with the mutex lock I omitted but no, adding the mutex lock to `ao_aaudio` or remove it from `ao_opensles` does not change anything.

**Comment by tomty89** [comment]
The segfault is most probably just a bug in libaaudio itself, as I can reproduce exactly the same behavior with the following program:
```
#include <stdio.h>
#include <aaudio/AAudio.h>

int main() {
  aaudio_result_t result;
  AAudioStreamBuilder *builder;
  AAudioStream *stream;

  result = AAudio_createStreamBuilder(&builder);
  printf("%d\n", result);
  result = AAudioStreamBuilder_openStream(builder, &stream);
  printf("%d\n", result);
  result = AAudioStream_close(stream);
  printf("%d\n", result);
  result = AAudioStreamBuilder_delete(builder);
  printf("%d\n", result);
  while (1)
    ;
}
```
`AAudioStream_close` always causes segfault upon app switch. If it is commented out, the program stays running well.

File: audio/out/ao_aaudio.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    AAudioStream_close(p->stream);
+    AAudioStreamBuilder_delete(p->builder);
+}
+
+static aaudio_data_callback_result_t buffer_callback(AAudioStream *stream, void *userData, void *audioData, int32_t numFrames) {
+    struct ao *ao = userData;
+    double delay;
+
+    //Copied from ao_opensles; what's the 2?
+    delay = 2 * numFrames / (double)ao->samplerate;
```
Comment (tmm1) [inline audio/out/ao_aaudio.c]: This doesn't apply to pull based drivers. You want to get rid of this, and use `AAudioStream_getTimestamp` to query the delay.

File: audio/out/ao_aaudio.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    ao_read_data(ao, &audioData, numFrames, mp_time_us() + 1000000LL * delay);
+
+    //Always?
```
Comment (tmm1) [inline audio/out/ao_aaudio.c]: Always returning success is fine. This callback will only be invoked while in a playing state.

**Pull request closed without merging**